DRA: refactor claim allocator

pohly opened the pull request:
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

The goal is to maintain different versions of the allocator logic. We already had one incidence where adding an alpha feature caused a regression also when it was disabled. Not everything can be implemented within obviously correct if
branches.    

This also opens the door for implementing different alternatives.

#### Which issue(s) this PR is related to:

N/A

#### Special notes for your reviewer:

The initial commit cleans up the API, the second just copies code around.

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

cc @mortent

pohly committed:
DRA allocator: export DeviceClassLister

It appears in the package API, so should be exported. For example,
an implementer of the interface might want to write a type assertion.

pohly committed:
DRA allocator: refactor code

The goal is to maintain different version of the allocator logic. We already
had one incidence where adding an alpha feature caused a regression also when
it was disabled. Not everything can be implemented within obviously correct if
branches.

This also opens the door for implementing different alternatives.

The code just gets moved around for now.

pohly committed:
DRA allocator: fork the code

The "incubating" variant is used only if additional features are needed.
At the moment, its code is identical to the "stable" variant except for
the different declaration of the supported features.

    $ staging/src/k8s.io/dynamic-resource-allocation/structured/internal/diff.sh
    *** staging/src/k8s.io/dynamic-resource-allocation/structured/internal/stable/allocator_stable.go	2025-07-10 15:40:13.370563544 +0200
    --- staging/src/k8s.io/dynamic-resource-allocation/structured/internal/incubating/allocator_incubating.go	2025-07-10 15:54:49.849262375 +0200
    ***************
    *** 44,57 ****
      	return internal.MakeDeviceID(driver, pool, device)
      }

    ! // SupportedFeatures does not include any additional features.
    ! // The "stable" implementation can only be used if all of those
    ! // are off.
    ! //
    ! // At least that's its purpose conceptually: in practice, all
    ! // code supporting the additional features is still there.
    ! // It could get removed.
    ! var SupportedFeatures = internal.Features{}

      type Allocator struct {
      	features         Features
    --- 44,58 ----
      	return internal.MakeDeviceID(driver, pool, device)
      }

    ! // SupportedFeatures includes all additional features,
    ! // making this the variant that is used when any of those
    ! // are enabled.
    ! var SupportedFeatures = internal.Features{
    ! 	AdminAccess:          true,
    ! 	PrioritizedList:      true,
    ! 	PartitionableDevices: true,
    ! 	DeviceTaints:         true,
    ! }

      type Allocator struct {
      	features         Features

Running the allocator tests against the top-level NewAllocator shows that it
branches into the different allocators as intended:

    === RUN   TestAllocator
    === RUN   TestAllocator/count-devices-some-allocated-admin-access
        allocator_incubating.go:116: I0710 15:56:24.673253] Starting allocation numClaims=1
    ...
    === RUN   TestAllocator/with-constraint-not-matching-string-attribute
        allocator_stable.go:115: I0710 15:56:24.703793] Starting allocation numClaims=1

mortent commented on staging/src/k8s.io/dynamic-resource-allocation/structured/internal/allocatortesting/allocator_testing.go:
```diff
@@ -0,0 +1,3544 @@
+			allocatedDevices := slices.Clone(tc.allocatedDevices)
+			slices := slices.Clone(tc.slices)
+
+			allocator, err := newAllocator(ctx, tc.features, unwrap(claimsToAllocate...), sets.New(allocatedDevices...), classLister, slices, cel.NewCache(1))
```
Can we compare the features supported by the allocator being tested and the features required by the test and just skip tests that are not supported by the allocator. That way, I think we should be able to mostly just add new tests to this file.

pohly replied:
Yes, that's the plan. I couldn't do it for your prioritized list improvement because it has not separate feature gate, but for consumable capacity that'll work. I can add the infrastructure now.

pohly committed:
DRA allocator: select tests based on supported features

This is how we can add tests for e.g. consumable capacity and then run those
tests only for a future "experimental" allocator.

At the moment, the "stable" allocator could run the tests and did before. With
this change, only tests for core functionality run, leading to some dead code
which (as in production!) does not get executed by the tests anymore.

Before:

$ go test -cover -count=1 ./staging/src/k8s.io/dynamic-resource-allocation/structured/...
ok  	k8s.io/dynamic-resource-allocation/structured	0.225s	coverage: 71.4% of statements
	k8s.io/dynamic-resource-allocation/structured/internal		coverage: 0.0% of statements
	k8s.io/dynamic-resource-allocation/structured/internal/allocatortesting		coverage: 0.0% of statements
ok  	k8s.io/dynamic-resource-allocation/structured/internal/incubating	0.217s	coverage: 92.1% of statements
ok  	k8s.io/dynamic-resource-allocation/structured/internal/stable	0.110s	coverage: 64.9% of statements

After:

$ go test -cover -count=1 ./staging/src/k8s.io/dynamic-resource-allocation/structured/...
ok  	k8s.io/dynamic-resource-allocation/structured	0.208s	coverage: 71.4% of statements
	k8s.io/dynamic-resource-allocation/structured/internal		coverage: 0.0% of statements
	k8s.io/dynamic-resource-allocation/structured/internal/allocatortesting		coverage: 0.0% of statements
ok  	k8s.io/dynamic-resource-allocation/structured/internal/incubating	0.208s	coverage: 92.1% of statements
ok  	k8s.io/dynamic-resource-allocation/structured/internal/stable	0.114s	coverage: 64.9% of statements

pohly commented:
@mortent: let's merge this now, then @sunya-ch can rebase her branch and add an "experimental" variant. Your change in https://github.com/kubernetes/kubernetes/pull/130593 goes into "internal/incubating/allocator_incubating.go".

@sunya-ch: you can start your branch with the last commit from https://github.com/pohly/kubernetes/pull/new/dra-allocator-consumable-capacity, then apply all of your changes to allocator.go to "internal/experimental/allocator_experimental.go".

I was wondering about how we will detect "indirect conflicts": suppose 130593 changes "allocator_incubating.go" after it was forked into "allocator_experimental.go". That breaks the "experimental is superset of incubating" principle, but it does not show up as a merge conflict because it's different files.

The answer is that unit tests for the change in 130593 will fail when applied to "experimental". We just need to be diligent about really having unit tests that fail without some change. Should be fine.

Editorial changes (like comment updates) won't fail that way, but they are also less critical. The "diff.sh" script will catch those, we just need to check manually.

mortent commented:
> @mortent: let's merge this now, then @sunya-ch can rebase her branch and add an "experimental" variant. Your change in #130593 goes into "internal/incubating/allocator_incubating.go".

Sounds good. Our PRs for mixins and device bindings also needs to be updated for this.

mortent commented:
/lgtm

But looks like there are some linter errors that need to be fixed.

pohly committed:
DRA allocator: fix linter hint

ST1016: methods on the same type should have the same receiver name (seen 10x "alloc", 1x "a") (staticcheck)

pohly commented:
> "methods on the same type should have the same receiver name" may be worth fixing. I'll probably do it.

Fixed.

pohly committed:
DRA allocator: skip unsupported test cases

Previously, they were not even started. Skipping them makes them visible in `go
test -v` output.

To simplify producing the list of unsupported features, conversion to a set
replaces the hand-written FeaturesAnd.

pohly commented on staging/src/k8s.io/dynamic-resource-allocation/structured/internal/allocatortesting/allocator_testing.go:
```diff
@@ -3486,6 +3490,11 @@ func TestAllocator(t *testing.T, newAllocator func(
 	for name, tc := range testcases {
+		if internal.FeaturesAnd(tc.features, supportedFeatures) != tc.features {
+			// Skip the test, at least one of its required features isn't supported
+			// and the test would fail.
```
I was trying to avoid increasing the number of tests which show up in testgrid, but as I now think about it again, that concerns is unnecessary: we already prune tests, so having more sub-tests won't make a difference.

To simplify this, I switched to using the sets package. I kept the struct because I feel that is is a bit easier to work with in the rest of the code.

nojnhuh reviewed (approved):
/lgtm

sanposhiho commented:
Only checked the interface + scheduler change, and lgtm. But, are you going to fix the linter reports?

pohly commented:
> Only checked the interface + scheduler change, and lgtm.

Please approve.

> But, are you going to fix the linter reports?

No, I am intentionally copying code around verbatim and decided against fixing that minor style issue.

pohly commented:
/skip

Let's make the red x go away :smile:

sanposhiho commented:
/approve
